### PR TITLE
Python 3.5 is dead -- long live Python 3.6!

### LINKED ISSUE #2266: Upgrade to Python 3.6+
This issue tracks all roadblocks we need to overcome to make use of Python 3.6-exclusive features. There's not immediate need to upgrade right away, but it's useful to have an overview of what's left nonetheless.

### Windows

✔️ Windows users can just install Python 3.6
✔️ Windows Subsystem for Linux currently contains Ubuntu 16.04, but a `do-release-upgrade` can be used without problems

### macOS

❌  High Sierra only has Python 2.7.10 and no Python 3
✔️ Homebrew is already on Python 3.6.3

### Ubuntu

✔️ Bionic 18.04 (next LTS release in April 2018) will have Python 3.6.3 (or a more recent patch-level)
✔️ Artful 17.10 has Python 3.6.3
❌ Zesty 17.04 has Python 3.5.3
❌ Ubuntu 16.04 LTS is still on 3.5.1

### Debian

✔️ testing: Debian 10 (buster) currently contains Python 3.6.3 -- no release date yet
❌ stable: Debian 9 (stretch) is still on 3.5.3

### Arch Linux

✔️ rolling release: has Python 3.6.3 as `python` package

### Fedora

✔️ Fedora 27 has Python 3.6.3
✔️ Fedora 26 has Python 3.6.3
❌ Fedora 25 has Python 3.5.4

### Kali Linux

✔️ kali-roll has Python 3.6.3 -- rolling release since January 2016
✔️ kali-dev has Python 3.6.3

### PyInstaller

- PyInstaller supports Python 3.6 since PyInstaller 3.3. We still have to upgrade to PyInstaller 3.3. Updating to PyInstaller 3.3.0 ends with a RecursionError within PyInstaller.

### PR DESCRIPTION
**Pull request by Kriechi**
fixes #2266

### CODE REVIEW

File: docs/src/content/overview-installation.md [hunk 1]
```diff
@@ -73,14 +73,14 @@ security considerations apply as for our binary packages.
...
-Please make sure to install Python 3.5 (or higher) and pip3 for your
+Please make sure to install Python 3.6 (or higher) and pip3 for your
 distribution. If your distribution does not provide a suitable Python
 version, you can use [pyenv](https://github.com/yyuu/pyenv) to get a
 recent Python environment.
 
 {{< highlight bash  >}}
-sudo apt install python3-pip # Debian 8 or higher, Ubuntu 16.04 or higher
-sudo dnf install python3-pip # Fedora 24 or higher
+sudo apt install python3-pip # Debian 10 or higher, Ubuntu 18.04 or higher
```
Comment (mhils) [inline docs/src/content/overview-installation.md]: Ubuntu 17.10 is good enough.

File: mitmproxy/tools/main.py [hunk 2]
```diff
@@ -2,11 +2,11 @@
 
 import sys
 
-if sys.version_info < (3, 5):
+if sys.version_info < (3, 6):
     # This must be before any mitmproxy imports, as they already break!
     # Keep all other imports below with the 'noqa' magic comment.
     print("#" * 49, file=sys.stderr)
-    print("# mitmproxy only supports Python 3.5 and above! #", file=sys.stderr)
+    print("# mitmproxy only supports Python 3.6! #", file=sys.stderr)
```
Comment (mhils) [inline mitmproxy/tools/main.py]: why skip the "and above"?

File: test/mitmproxy/utils/test_typecheck.py [hunk 3]
```diff
@@ -32,10 +32,9 @@ def test_check_union():
     with pytest.raises(TypeError):
         typecheck.check_option_type("foo", [], typing.Union[int, str])
 
-    # Python 3.5 only defines __union_params__
     m = mock.Mock()
     m.__str__ = lambda self: "typing.Union"
-    m.__union_params__ = (int,)
+    m.__args__ = (int,)
```
Comment (mhils) [inline test/mitmproxy/utils/test_typecheck.py]: The whole mock is not required anymore, that was just to test 3.5. (same below)

File: tox.ini [hunk 4]
```diff
@@ -21,7 +21,7 @@ commands =
 commands =
   mitmdump --version
   flake8 --jobs 8 mitmproxy pathod examples test release
-  python test/filename_matching.py
+  python3 ./test/filename_matching.py
```
Comment (mhils) [inline tox.ini]: IIRC this breaks somewhere. tox normalizes to "python" AFAIK.
(also below)

File: mitmproxy/tools/main.py (hunk 2 above)
Comment (Kriechi) [reply mitmproxy/tools/main.py]: Unless we start testing for 3.7 I think we should not claim that it works for "and above".
Felt kinda wrong, semver and everything... But I can revert it if you have strong feeling about it.

File: tox.ini (hunk 4 above)
Comment (Kriechi) [reply tox.ini]: works for me and Travis .... but not Appveyor.  ¯\_(ツ)_/¯

File: test/mitmproxy/utils/test_typecheck.py (hunk 3 above)
Comment (Kriechi) [reply test/mitmproxy/utils/test_typecheck.py]: I fiddled with it a bit, without much success... 
What do we have to set `m` to, to make it work?

File: mitmproxy/tools/main.py (hunk 2 above)
Comment (mhils) [reply mitmproxy/tools/main.py]: Yes - once 3.7 is supposed to be backward compatible, we have no particular tie-ins to 3.6.

File: test/mitmproxy/utils/test_typecheck.py (hunk 3 above)
Comment (mhils) [reply test/mitmproxy/utils/test_typecheck.py]: Just remove the whole mock part entirely, this was just to get test coverage without running the old version.

**Pull request merged**